### PR TITLE
Con-4276 search completed assessments

### PR DESCRIPTION
**Pull request by kola-samuel**

### CODE REVIEW

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs [hunk 1]
```diff
@@ -300,40 +300,129 @@ public async Task UpdateCertificatesReadyToPrintInBatch(Guid[] certificateIds, i
...
+                int pageIndex,
+                int pageSize,
+                string searchTerm,
+                string sortColumn,
+                bool sortDescending,
+                List<string> statuses)
+        {
+
+            var certificatesQuery = (from certificate in _context.Certificates
+                 .Include(q => q.Organisation)
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Data/CertificateRepository.cs:315]: Why is Organisation included here and then also joined at line 318? and wouldn't it read better to just use all join syntax statements?

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs [hunk 2]
```diff
@@ -300,40 +300,129 @@ public async Task UpdateCertificatesReadyToPrintInBatch(Guid[] certificateIds, i
...
+                            .OrderBy(x => x.Certificate.CreatedAt);
+                    }
+
+                    break;
+            }
+
+            return certificatesQuery;
+        }
+
         public async Task<int> GetCertificatesCount(string endPointAssessorOrganisationId, List<string> statuses)
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Data/CertificateRepository.cs:418]: This looks like it was a count of what the get certificates used to do - is it still used?

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs [hunk 3]
```diff
@@ -573,4 +662,10 @@ [Certificates] c
                    transaction: _unitOfWork.Transaction);
         }
     }
+    class SortableCertificateInfo
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Data/CertificateRepository.cs:665]: New line

File: src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs [hunk 4]
```diff
@@ -13,40 +16,110 @@
...
+
+        [HttpGet(nameof(ChangePageCertificateHistory))]
+        public async Task<IActionResult> ChangePageCertificateHistory(int pageIndex = DefaultPageIndex)
+        {
+            _certificateHistorySession.CertificateHistoryPageIndex = pageIndex;
+
+            var vm = await MapViewModelFromSession();
+            return View(nameof(Index), vm);
+        }
+        private void UpdateCertificateHistorySortDirection(string sortColumnName, string sortDirection)
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs:82]: Add some white space between method

File: src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs [hunk 5]
```diff
@@ -13,40 +16,110 @@
...
+                    viewModel.SortColumn.ToString(),
+                    viewModel.SortDirection == "Asc" ? 1 : 0);
 
-            return View("Index", model);
+            return viewModel;
+        }
+
+        private void SetDefaultSession()
+        {
+            _certificateHistorySession.CertificateHistorySearchTerm = string.Empty; ;
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs:119]: Remove the ;

File: src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs [hunk 6]
```diff
@@ -13,40 +16,110 @@
...
+            return RedirectToAction(nameof(Index));
+        }
+
+        [HttpGet(nameof(Search))]
+        public async Task<IActionResult> Search(string searchTerm)
+        {
+            _certificateHistorySession.CertificateHistorySearchTerm = searchTerm ?? string.Empty;
+
+            // reset the page indexes as the new results may have less pages
+            _certificateHistorySession.CertificateHistoryPageIndex = 1;
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs:61]: There is a DefaultPageIndex you could use

File: src/SFA.DAS.AssessorService.Web/Resources/Views/CertificateHistory/Index.en.resx [hunk 7]
```diff
@@ -195,4 +195,41 @@
...
+  <data name="LatestCertificateDetails" xml:space="preserve">
+    <value>Latest Certificate Details</value>
+  </data>
+  <data name="DateAndTime" xml:space="preserve">
+    <value>Date and time</value>
+  </data>
+  <data name="Status" xml:space="preserve">
+    <value>Status</value>
+  </data>
+  <data name="DateSubmitted" xml:space="preserve">
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Web/Resources/Views/CertificateHistory/Index.en.resx:213]: What are these used for - there is a DisplayName method about which looks similar?

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.AssessorService.Web/Resources/Views/CertificateHistory/Index.en.resx (hunk 7 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Web/Resources/Views/CertificateHistory/Index.en.resx]: Removed

File: src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs (hunk 6 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs]: Changed

File: src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs (hunk 5 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs]: done

File: src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs (hunk 4 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Web/Controllers/CertificateHistoryController.cs:84]: There is a line spacing

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs (hunk 3 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Data/CertificateRepository.cs]: Done

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs (hunk 2 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Data/CertificateRepository.cs]: Removed

**Review by ChrisJWoodcock — APPROVED** [review]

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs (hunk 1 above)
Comment (cofaulco) [reply src/SFA.DAS.AssessorService.Data/CertificateRepository.cs]: coming back to this one, does EF not generate the joins based on the Include()? Could this not all be achieved with a LINQ where and  move the JSON serialization within the Select()?

`var certificatesQuery = _context.Certificates
                .Include(c => c.Organisation)
                .Include(c => c.CertificateLogs)
                .Include(c => c.CertificateBatchLog)
                .Where(c => c.Organisation.EndPointAssessorOrganisationId == endPointAssessorOrganisationId && !statuses.Contains(c.Status))
                .Select(c => new SortableCertificateInfo
                {
                    Certificate = c,
                    Organisation = c.Organisation,
                    CertificateData = JsonConvert.DeserializeObject<CertificateData>(c.CertificateData)
                });`

File: src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/CertificateSummaryResponse.cs [hunk 8]
```diff
@@ -51,4 +50,4 @@ public class CertificateStatusList
         public DateTime CertificateDate { get; set; }
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/CertificateSummaryResponse.cs:50]: The CertificateStatusList remains here - is it required? this seems to have been removed almost completely, I think it was never in the AC's has it just been dropped as a mistake?

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/CertificateSummaryResponse.cs (hunk 8 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Api.Types/Models/Certificates/CertificateSummaryResponse.cs]: removed

File: src/SFA.DAS.AssessorService.Data/CertificateRepository.cs (hunk 1 above)
Comment (kola-samuel) [reply src/SFA.DAS.AssessorService.Data/CertificateRepository.cs]: Yes but I have a more simplex version with  out .Deserialization out of SQL

**Review by ChrisJWoodcock — APPROVED** [review]

**Review by ChrisJWoodcock — APPROVED** [review]

**Pull request merged**